// testcafe/lib/shared/utils/event-emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore
const hammerhead_1 = require("../../client/driver/deps/hammerhead");
class EventEmitter {
    constructor() {
        this._eventsListeners = {};
    }
    on(evt, listener) {
        if (!this._eventsListeners[evt])
            this._eventsListeners[evt] = [];
        this._eventsListeners[evt].push(listener);
    }
    once(evt, listener) {
        this.on(evt, (...args) => {
            this.off(evt, listener);
            return listener(...args);
        });
    }
    off(evt, listener) {
        const listeners = this._eventsListeners[evt];
        if (listeners)
            this._eventsListeners[evt] = hammerhead_1.nativeMethods.arrayFilter.call(listeners, (item) => item !== listener);
    }
    offAll(evt) {
        if (evt)
            this._eventsListeners[evt] = [];
        else
            this._eventsListeners = {};
    }
    emit(evt, ...args) {
        const listeners = this._eventsListeners[evt];
        if (!listeners)
            return;
        for (let i = 0; i < listeners.length; i++) {
            try {
                listeners[i].apply(this, args);
            }
            catch (e) {
                // Hack for IE: after document.write calling IFrameSandbox event handlers
                // rises 'Can't execute code from a freed script' exception because document has been
                // recreated
                if (e.message && e.message.indexOf('freed script') > -1)
                    this.off(evt, listeners[i]);
                else
                    throw e;
            }
        }
    }
}
exports.default = EventEmitter;
module.exports = exports.default;
