// testcafe/lib/shared/actions/cursor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axis_values_1 = __importDefault(require("../utils/values/axis-values"));
// @ts-ignore
const hammerhead_1 = require("../../client/driver/deps/hammerhead");
class Cursor {
    constructor(activeWin, ui) {
        this._ui = ui;
        // NOTE: the default position should be outside of the page (GH-794)
        this._x = -1;
        this._y = -1;
        this._activeWindow = activeWin;
    }
    _ensureActiveWindow(win) {
        if (this._activeWindow === win || this._activeWindow === win.parent)
            return;
        if (this._activeWindow.parent !== win)
            this._activeWindow = win;
    }
    isActive(currWin) {
        this._ensureActiveWindow(currWin);
        return this._activeWindow === currWin;
    }
    setActiveWindow(win) {
        this._activeWindow = win;
    }
    getActiveWindow(currWin) {
        this._ensureActiveWindow(currWin);
        return this._activeWindow;
    }
    getPosition() {
        return new axis_values_1.default(this._x, this._y);
    }
    move(point) {
        this._x = point.x;
        this._y = point.y;
        return this._ui.move(point);
    }
    hide() {
        if (this._ui.hide)
            return this._ui.hide();
        return hammerhead_1.Promise.resolve();
    }
    show() {
        if (this._ui.show)
            return this._ui.show();
        return hammerhead_1.Promise.resolve();
    }
    leftButtonDown() {
        return this._ui.leftButtonDown();
    }
    rightButtonDown() {
        return this._ui.rightButtonDown();
    }
    buttonUp() {
        return this._ui.buttonUp();
    }
}
exports.default = Cursor;
module.exports = exports.default;
