// testcafe/lib/shared/actions/automations/click/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const adapter_1 = require("../../../adapter");
const visible_element_automation_1 = __importDefault(require("../visible-element-automation"));
const delay_1 = __importDefault(require("../../../utils/delay"));
// @ts-ignore
const hammerhead_1 = require("../../../../client/automation/deps/hammerhead");
class ClickAutomation extends visible_element_automation_1.default {
    constructor(element, clickOptions, win, cursor) {
        super(element, clickOptions, win, cursor);
        this.modifiers = clickOptions.modifiers;
        this.strategy = adapter_1.adapter.automations.click.createMouseClickStrategy(this.element, clickOptions.caretPos);
    }
    _mousedown(eventArgs) {
        return this.strategy.mousedown(eventArgs);
    }
    _mouseup(element, eventArgs) {
        return this.strategy.mouseup(element, eventArgs);
    }
    run(useStrictElementCheck) {
        let eventArgs;
        return this
            ._ensureElement(useStrictElementCheck)
            .then(({ element, clientPoint, screenPoint, devicePoint }) => {
            eventArgs = {
                point: clientPoint,
                screenPoint: screenPoint,
                element: element,
                options: hammerhead_1.utils.extend({
                    clientX: clientPoint === null || clientPoint === void 0 ? void 0 : clientPoint.x,
                    clientY: clientPoint === null || clientPoint === void 0 ? void 0 : clientPoint.y,
                    screenX: devicePoint === null || devicePoint === void 0 ? void 0 : devicePoint.x,
                    screenY: devicePoint === null || devicePoint === void 0 ? void 0 : devicePoint.y,
                }, this.modifiers),
            };
            // NOTE: we should raise mouseup event with 'mouseActionStepDelay' after we trigger
            // mousedown event regardless of how long mousedown event handlers were executing
            return Promise.all([delay_1.default(this.automationSettings.mouseActionStepDelay), this.cursor
                    .leftButtonDown()
                    .then(() => this._mousedown(eventArgs)),
            ]);
        })
            .then(() => this.cursor.buttonUp())
            .then(() => this._getElementForEvent(eventArgs))
            .then(element => {
            return element ? this._mouseup(element, eventArgs) : null;
        });
    }
}
exports.default = ClickAutomation;
module.exports = exports.default;
